File: sections/63_atcoder_difficulty_ladder/problems/38_anticube/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, minimum_length, maximum_length;
    if (!(cin >> n >> minimum_length >> maximum_length)) return 0;
    vector<long long> prefix(n + 1);
    for (int index = 1; index <= n; ++index) {
        long long value;
        cin >> value;
        prefix[index] = prefix[index - 1] + value;
    }

    deque<int> candidates;
    long long answer = numeric_limits<long long>::lowest();
    for (int right = 1; right <= n; ++right) {
        int entering = right - minimum_length;
        if (entering >= 0) {
            while (!candidates.empty() && prefix[candidates.back()] >= prefix[entering]) {
                candidates.pop_back();
            }
            candidates.push_back(entering);
        }
        while (!candidates.empty() && candidates.front() < right - maximum_length) {
            candidates.pop_front();
        }
        if (!candidates.empty()) {
            answer = max(answer, prefix[right] - prefix[candidates.front()]);
        }
    }
    cout << answer << '\n';
}

File: sections/63_atcoder_difficulty_ladder/problems/38_anticube/solution.py
import sys
from collections import deque


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    n, minimum_length, maximum_length = data[:3]
    values = data[3:]
    prefix = [0] * (n + 1)
    for index, value in enumerate(values, 1):
        prefix[index] = prefix[index - 1] + value

    candidates: deque[int] = deque()
    answer = -10**30
    for right in range(1, n + 1):
        entering = right - minimum_length
        if entering >= 0:
            while candidates and prefix[candidates[-1]] >= prefix[entering]:
                candidates.pop()
            candidates.append(entering)
        while candidates and candidates[0] < right - maximum_length:
            candidates.popleft()
        if candidates:
            answer = max(answer, prefix[right] - prefix[candidates[0]])

    print(answer)


if __name__ == "__main__":
    main()
